text/1184415/sync_hirschberg_sinclair_second_edition.tex

```diff
@@ -0,0 +1,166 @@
+\(\mathcal{O}(n \log n)\) messages.
+Previous successes belong to:
+\begin{itemize}
+    \item LeLann [2] algorithm, wich requires $O(n^2)$ messages.
+    \item Chang and Roberts [3]: average - O(n log n), but worst case still \(\mathcal{O}(n^2)\)
+\end{itemize}
+
+
+\section*{Algorithm}
+A naive algorithm, which achieves
+\(\mathcal{O}(n^2)\), involves simply passing our
+$Id$ in different directions. However, this quadratic time complexity can be reduced to \(\mathcal{O}(n)\) rounds. Our algorithm takes steps of length
+$2^i$, during which the number of active processes is reduced. Those that fall within the range $2^i$ from a more possible candidate for the role of leader will become inactive, resulting in a significant reduction in the number of messages sent.\\
+"Must have" to be in head:
+\begin{enumerate}
+    \item \(sendpass\) - sending message in same direction as recieved was going.
+    \item \(sendecho\) - returns message back to sender of recieved message.
+\end{enumerate}
+
+Lets take a look on pseudocode from [1].
+\begin{lstlisting}
+The Algorithm
+
+/// if after coming back process's messages it still has
+/// "candidate" status, it means, that it is the biggest in
+/// range 2^i around him
+run_candidate:
+    status <--"candidate"
+    maxnum <-- 1
+    WHILE status = "candidate"
+        sendboth ("from", myvalue, 0, maxnum)
+        await both replies (but react to other messages)
+        IF either reply is "no" THEN status <-- "lost"
+        maxnum <-- 2*maxnum
+
+
+/// the function thanks to wich candidates eleminate number of possible canditates
+/// and have abbility to get info, that they have won
+receive_message ("from", value, num, maxnum):
+    IF value < myvalue
+        sendecho ("no", value)
+    IF value > myvalue
+        status <-- "lost"
+        num <- num + 1
+        IF num < maxnum
+            sendpass ("from", value, num, maxnum)
+        ELSE
+            sendeeho ("ok", value)
+    IF value = myvalue
+        status <- "won"
+
+
+///besides accepting his own message, which came back to him,
+///the processor can play the role of a transit, inactive node
+receive_message ("no", value) or ("ok", value)
+    IF value != myvalue
+        senpass the message
+    ELSE
+        this is a reply the processor was awaiting
+
+///one from me, just for simle and correct ending
+///of this election, eleminating all possible unread messages
+receive_message ("end", value):
+    inform next processor of end
+    stop process
+\end{lstlisting}
+
+This pseudocode is more suitable for a solution based on maintaining references to neighboring processes. However, in the implementation I provided, solution stuck to to passing information through channels, requiring a slight modification in the code structure. Nevertheless the given pseudocode effectively reflects the algorithm's flow, as much, as it reduces the number of messages).
+
```

Comment:
I suppose this could become irrelevant when you migrate to graph and synchronizer as it's implemented in the library.

Reply:
Has been removed